src/zenml/utils/metadata_utils.py

```diff
@@ -114,150 +136,152 @@ def log_metadata(
+                "combination of `model_name` and `model_version`."
             )
-        else:
-            step_context = None
-            with contextlib.suppress(RuntimeError):
-                step_context = get_step_context()
+        mv = step_context.model
+        mv.log_metadata(metadata)
 
-            if step_context:
-                step_context.add_output_metadata(
-                    metadata=metadata, output_name=artifact_name
-                )
-            else:
-                artifact_version_model = client.get_artifact_version(
-                    name_id_or_prefix=artifact_name
-                )
-                client.create_run_metadata(
-                    metadata=metadata,
-                    resource_id=artifact_version_model.id,
-                    resource_type=MetadataResourceTypes.ARTIFACT_VERSION,
+    # Log metadata to an artifact version by its name and version
+    elif artifact_name is not None and artifact_version is not None:
+        artifact_version_model = client.get_artifact_version(
+            name_id_or_prefix=artifact_name, version=artifact_version
+        )
+        client.create_run_metadata(
+            metadata=metadata,
+            resources=[
+                RunMetadataResource(
+                    id=artifact_version_model.id,
+                    type=MetadataResourceTypes.ARTIFACT_VERSION,
                 )
+            ],
+        )
 
-    # If the user directly provides an artifact_version_id, we use the client to
-    # fetch is and attach the metadata accordingly.
+    # Log metadata to an artifact version by its ID
     elif artifact_version_id is not None:
-        artifact_version_model = client.get_artifact_version(
-            name_id_or_prefix=artifact_version_id,
-        )
         client.create_run_metadata(
             metadata=metadata,
-            resource_id=artifact_version_model.id,
-            resource_type=MetadataResourceTypes.ARTIFACT_VERSION,
+            resources=[
+                RunMetadataResource(
+                    id=artifact_version_id,
+                    type=MetadataResourceTypes.ARTIFACT_VERSION,
+                )
+            ],
+        )
+
+    # Log metadata to an artifact version through the step context
+    elif infer_artifact is True:
+        try:
+            step_context = get_step_context()
+        except RuntimeError:
+            raise ValueError(
+                "When you are using the `infer_artifact` option when you call "
+                "`log_metadata`, it must be called inside a step with outputs."
+                "Otherwise, you can provide a `artifact_version_id` or a "
+                "combination of `artifact_name` and `artifact_version`."
+            )
+
+        step_output_names = list(step_context._outputs.keys())
+
+        if artifact_name is not None:
+            # If a name provided, ensure it is in the outputs
+            assert artifact_name in step_output_names, (
```

Comment:
This should be a ValueError, assert is not something that should be used to verify conditions

Reply:
Changed it now.